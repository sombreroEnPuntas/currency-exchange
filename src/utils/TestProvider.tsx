import React from 'react'
import { Provider } from 'react-redux'

import store from '../data'

const TestProvider = ({ children }) => (
  <Provider store={store}>{children}</Provider>
)

export default TestProvider
