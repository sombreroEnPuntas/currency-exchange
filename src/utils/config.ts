export default {
  exchangeRateAPIRoute: 'https://openexchangerates.org/api/latest.json',
  availableCurrencies: ['EUR', 'GBP', 'USD'],
}
