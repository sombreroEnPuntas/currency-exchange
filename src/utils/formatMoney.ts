import { Currency } from '../types'

const roundToTwoDecimals = (number: number) =>
  Math.round((number + Number.EPSILON) * 100) / 100

const localFormatter = (currency?: Currency) =>
  new Intl.NumberFormat('en-US', { style: 'currency', currency })

/**
 * Formats a monetary amount as currency when [currency] is given,
 * or two decimals otherwise.
 *
 * @param {number} value monetary amount
 * @param {Currency} [currency] ISO currencies
 */
const formatMoney = (value: number, currency?: Currency) =>
  currency
    ? localFormatter(currency).format(roundToTwoDecimals(value))
    : roundToTwoDecimals(value).toFixed(2)

export default formatMoney
