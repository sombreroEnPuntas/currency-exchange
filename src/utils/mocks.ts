import { Pocket } from '../data/pockets'

export const pocketsMock: Pocket[] = [
  { availableAmount: 58.33, currency: 'GBP' },
  { availableAmount: 25.51, currency: 'USD' },
  { availableAmount: 0, currency: 'EUR' },
]
