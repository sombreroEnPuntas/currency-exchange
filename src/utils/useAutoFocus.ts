import { useRef, useEffect } from 'react'

const useAutoFocus = () => {
  const htmlElRef = useRef(null)

  useEffect(() => {
    htmlElRef.current && htmlElRef.current.focus()
  }, [htmlElRef])

  return htmlElRef
}

export default useAutoFocus
