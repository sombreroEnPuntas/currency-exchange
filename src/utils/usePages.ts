import { useEffect, useState } from 'react'

interface Deps {
  list: any[]
  pageSize: number
}
const usePages = ({ list, pageSize }: Deps) => {
  const [cursor, setCursor] = useState(0)
  const [pages, setPages] = useState([])

  useEffect(() => {
    setPages(
      list.reduce((all, page, index) => {
        const currentPage = Math.floor(index / pageSize)
        all[currentPage] = [].concat(all[currentPage] || [], page)
        return all
      }, [])
    )
  }, [list, pageSize])

  return {
    handleBack: () => setCursor(cursor - 1 < 0 ? pages.length - 1 : cursor - 1),
    handleNext: () => setCursor(cursor + 1 > pages.length - 1 ? 0 : cursor + 1),
    page: pages[cursor] || [],
    pageNumber: cursor + 1,
    totalPages: pages.length,
  }
}

export default usePages
