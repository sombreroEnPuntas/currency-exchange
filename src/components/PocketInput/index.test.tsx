import React from 'react'
import { render } from '@testing-library/react'

// Tested Unit
import Component from '.'

// Utils
import { pocketsMock } from '../../utils/mocks'

// Mocks
const testPocket = pocketsMock.find(({ currency }) => currency === 'GBP')

// Mock data
const getProps = (customProps) => ({
  name: 'from',
  pocket: testPocket,
  ...customProps,
})

const setup = (customProps = {}) => {
  const utils = render(<Component {...getProps(customProps)} />)
  const inputEl = utils.getByLabelText('GBP')

  return { inputEl, ...utils }
}

describe('PocketInput', () => {
  it('Renders mandatory data', () => {
    const { container, getByText, inputEl } = setup()

    expect(inputEl).toHaveAttribute('id', 'from')
    expect(inputEl).toHaveAttribute('name', 'from')
    expect(inputEl).toHaveAttribute('type', 'number')

    expect(getByText(/You have Â£58.33/)).toBeVisible()

    expect(container).toMatchSnapshot()
  })

  it('Passes down rest props to input element', () => {
    const { inputEl } = setup({ required: true })

    expect(inputEl).toHaveAttribute('required')
  })
})
