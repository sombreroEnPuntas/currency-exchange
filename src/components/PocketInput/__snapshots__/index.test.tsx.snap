// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PocketInput Renders mandatory data 1`] = `
<div>
  <label
    for="from"
  >
    GBP
  </label>
  <input
    id="from"
    max="58.33"
    min="0"
    name="from"
    step="0.01"
    type="number"
  />
  <div>
    You have Â£58.33
  </div>
</div>
`;
